### ISSUE TITLE
Publish build artifacts

### ISSUE THREAD

**Issue by jplaisted** [linkedin/datahub#1804]
We should be publishing our build artifacts (jars, etc) so people don't need to build locally and link against those.

We should do this for non demo libraries and applications. Libraries are good for those that want to build on top of the code we've written (e.g. new metadata crawler that emits MCEs; a separate GMS; etc). Applications are good for those that want to customize their deployments without needing to build source code.

From https://github.com/linkedin/datahub/issues/1743#issuecomment-673011663

**Comment by mars-lan** [comment]
Need to be careful about what got published has real value. Once it's published we'll need to support it forever.

**Comment by jplaisted** [comment]
There should be some kind of EoL process for these. Plus we're still not on 1.0 yet...

**Comment by shakti-garg-saxo** [comment]
I understand the concerns around EoL process for it. 
My justification for request to publish metadata-models jar is that the bindings and avsc files there define the input contracts(MetadataChangeEvent) for the datahub. These contracts can be used by any external service for programmatic extension, rather than relying on internal tools which are build locally.

Happy to discuss more.

**Comment by mars-lan** [comment]
@shakti-garg-saxo completely agree with the reason. We're in the process of decoupling dependency on models from codes so that the models jar can be loaded almost like a plugin. It may still take a bit more work but we're not too far from that ideal state.

**Comment by jplaisted** [comment]
So just to be clear on our plans and order of operations here:

- We want to split out GMA in its own repository. We'll publish jars for this so DataHub can use them (anyone can, but point is we need to for DataHub :p ). This, for now, will include the DAOs, and maybe some testing utilities. Ideally in the future it would include the MCE/MAE jobs, but they are unfortunately coupled to the existing models.
  - Splitting this into a separate repository has some advantages. It helps make it a little more distinct from DataHub (you can use it without DataHub). It also helps us clarify its roadmap separately. GMA can probably hit 1.0 separately from DataHub.
  - Allows us to make some breaking changes without immediately needing to refactor DataHub. We don't want to make breaking changes often, but we'll probably have some set of breaking changes we want to get into 1.0.0. Breaking changes are going to be even more infrequent and carefully staged after 1.0.0.
  - Should be done before the end of September.
- Look into publishing library jars in DataHub (models, builders, other utilities).
- Look into publishing application jars in DataHub (frontend, gms, jobs).
- Look into decoupling more things, i.e. jobs from models, maybe make GMSes actually microservices as we suggest, see if we can even decouple the frontend... these are long term things and probably shouldn't block us from publishing our jars, imo. These would either be new projects/jars or new major version bumps :p

**Comment by mars-lan** [comment]
SGTM

**Comment by shakti-garg-saxo** [comment]
Great. It looks like a very good plan. 
Thanks.

**Comment by jplaisted** [comment]
As an fyi, we moved the backend (GMA / Generalized Metadata Architecture) to its own repo, and it publishes artifacts to bintray.

https://github.com/linkedin/datahub-gma

**Comment by shakti-garg** [comment]
Thanks John! Saw the changes in (https://github.com/linkedin/datahub/pull/1955).
Am i correct in understanding that datahub-gma will have separate release lifecycle than datahub?

**Comment by jplaisted** [comment]
Yes. Ideally we'll set up a continuous release for it soon, so ever push to master will publish a new version. But the versioning scheme will also be independent.

**Comment by jjoyce0510** [comment]
@jplaisted Hi John. I know we've gotten through the splitting of GMA out (bullet one). Is there a plan for the following work? That is, publishing jars for metadata models, builders, etc. as part of the release cycle?

**Comment by jplaisted** [comment]
I think this stalled a bit due to questions around model federation and the state of the DataHub repo. I'll let the community guide this discussion from here; should we publish artifacts today or wait for some potential larger refactor that may or may not happen?

**Comment by maggiehays** [comment]
We are now managing feature requests through our [new portal](https://feature-requests.datahubproject.io) to improve visibility & prioritize work into our public roadmap. If this request is still relevant, please create a new issue so we can gauge current interest from the Community & prioritize work accordingly.

We'll close out this issue in the meantime. Feel free to [reach out to me on DataHub Slack](https://slack.datahubproject.io/) with any questions!
